### PR TITLE
New MaskSet.get_cmap() implementation

### LINKED ISSUE #1011: Provide color map for mask set with `flag_values`
**Is your feature request related to a problem? Please describe.**

For dataset variables that provide a `flag_values` attribute, it would be convenient to get a matching matplotlib color mapping. For some variables we also provide `flag_colors` that should be used if so.

**Describe the solution you'd like**

Add method `get_cmap()` to `xcube.core.maskset.MaskSet`.

**Additional context**

CCI Land Cover datasets, `lccs_class` variable.

### PR DESCRIPTION
**Pull request by forman**
Closes #1011

@konstntokas please also test with original LC data in last webinar notebook.

Checklist:

* [x] Add unit tests and/or doctests in docstrings
* [x] Add docstrings and API docs for any new/modified user-facing classes and functions
* [x] New/modified features documented in `docs/source/*`
* [x] Changes documented in `CHANGES.md`
* [ ] GitHub CI passes
* [ ] AppVeyor CI passes
* [x] Test coverage remains or increases (target 100%)

### CODE REVIEW

File: xcube/core/maskset.py [hunk 1]
```diff
@@ -274,3 +302,31 @@ def _convert_flag_var_attribute_value(attr_value, attr_name):
         return np.array(attr_value)
 
     raise TypeError(f"attribute {attr_name!r} must be an integer array")
+
+
+def _sanitize_flag_values(
+    flag_var: xr.DataArray,
+    flag_values: np.ndarray,
+) -> Optional[np.ndarray]:
```
Comment (konstntokas) [inline xcube/core/maskset.py:308]: Why Optional?

**Review by konstntokas — CHANGES_REQUESTED** [review]
We still have the same problem that the user will need to add vmin and vmax manually. I find this, as a user, more confusing then getting vmin and vmax from the `get_cmap` method. See examples below: 

```
land_cover_mask = MaskSet(land_cover_2019.lccs_class)
cmap = land_cover_mask.get_cmap()
land_cover_2019.lccs_class.plot(cmap=cmap)
```
landcover

```
land_cover_mask = MaskSet(land_cover_2019.lccs_class)
cmap = land_cover_mask.get_cmap()
land_cover_2019.lccs_class.plot(cmap=cmap, vmin=0, vmax=220)
plt.show()
```
landcover_vmin_vax

The colormap is not really nice, but this method at least buts the values on the correct position. I did something like: 

```
def get_cmap(maskset):
    vals = maskset._flag_values
    extent = vals.max() - vals.min()
    vals_norm = (vals - vals.min()) / extent
    colors = [(0, 0, 0, 0)] + maskset._flag_colors
    color_mapping = [(v, c) for v, c in zip(vals_norm, colors)]
    for v, c in zip(vals, colors):
        print(v, c)
    return matplotlib.colors.LinearSegmentedColormap.from_list(
        str(maskset._flag_var.name), color_mapping, N=len(vals)
    )

land_cover_mask = MaskSet(land_cover_2019.lccs_class)
cmap = get_cmap(land_cover_mask)
land_cover_2019.lccs_class.plot(cmap=cmap, vmin=0, vmax=220)
plt.show()
```
landcover_oldcmap

This gives an aesthetically nicer colormap but distorts the actual values of course. I am indecisive whats better. 

Still we should do something about vmin and vmax, I think. Otherwise a lot of users will run into problems. Would it be possible to add a plotting routine to `MaskSet`?

**Comment by forman** [comment]
> We still have the same problem that the user will need to add vmin and vmax manually. 

Then, xarray is not calling the color map in the right way. The new color map is a [`matplotlib.colors.ListedColormap`](https://matplotlib.org/stable/api/_as_gen/matplotlib.colors.ListedColormap.html) and it's documentation (and ensured by our unit tests) says:

_This may be most useful when indexing directly into a colormap_

> I find this, as a user, more confusing then getting vmin and vmax from the `get_cmap` method.

That's right!

**Comment by forman** [comment]
> The colormap is not really nice, but this method at least buts the values on the correct position.

> This gives an aesthetically nicer colormap but distorts the actual values of course. I am indecisive whats better.

The Land Cover color mapping is not about aesthetics. The ultimate goal is to have an exact match between flag values and assigned color. It is really a pity that the provided `ListedColormap` is not used correctly by xarray. I'll need to update the implementation and also return a `(cmap, norm)` pair then, similar to your code.

File: xcube/core/maskset.py (hunk 1 above)
Comment (forman) [reply xcube/core/maskset.py:308]: Forgot to remove.

**Comment by forman** [comment]
> Still we should do something about vmin and vmax, I think. 

Yes, return a `(cmap, norm)` pair, see above.

> Would it be possible to add a plotting routine to `MaskSet`?

Yes, but let's keep it simple for time being. I will change impl. to return the pair.

**Comment by konstntokas** [comment]
Including matplotlib.colors.BoundaryNorm the colormap looks like this. The numbers are of course confusing. We would need to create the plot and assign the ticks to the center of each discrete color and assign the tick label to either the flag value or even flag name. 

landcover_bound_norm

File: xcube/core/maskset.py [hunk 2]
```diff
@@ -191,34 +194,67 @@ def get_mask(self, flag_name: str):
...
         Args:
             default: Default color map name in case a color mapping
-                cannot be created, e.g., ``flag_values`` are not defined.
+                cannot be created, e.g., ``flag_values`` are not defined or
+                if the flag values are not in the range [0, 2**16 - 1).
 
         Returns:
-            An suitable instance of ```matplotlib.colors.Colormap```
+            An suitable instance of ```matplotlib.colors.Colormap``` and
+            the corresponding ```matplotlib.colors.BoundaryNorm``` if applicable
```
Comment (forman) [inline xcube/core/maskset.py:209]:
```suggestion
            An suitable instance of ``matplotlib.colors.Colormap`` and
            the corresponding ``matplotlib.colors.BoundaryNorm`` if applicable
```

File: xcube/core/maskset.py [hunk 3]
```diff
@@ -191,34 +194,67 @@ def get_mask(self, flag_name: str):
...
-            return matplotlib.colors.LinearSegmentedColormap.from_list(
-                str(self._flag_var.name), colors
-            )
-        return matplotlib.colormaps.get_cmap(default)
+            flag_value_min = flag_values.min()
+            flag_value_max = flag_values.max()
+
+            if flag_value_min >= 0 and flag_value_max < _UINT16_MAX:
+                flag_colors = self._flag_colors
+                levels = np.append(flag_values - 0.5, flag_values[-1] + 0.5)
```
Comment (forman) [inline xcube/core/maskset.py:221]: Why 0.5 shift?

**Review by forman — COMMENTED** [review]
See suggestions

**Comment by konstntokas** [comment]
The plot now looks like this: 

Code snippet 

```
land_cover_mask = MaskSet(land_cover_2019.lccs_class)
cmap, norm = land_cover_mask.get_cmap()
land_cover_2019.lccs_class.plot.imshow(cmap=cmap, norm=norm)
plt.show()
```

landcover_bound_norm

File: xcube/core/maskset.py (hunk 3 above)
Comment (konstntokas) [reply xcube/core/maskset.py]: shift is removed, see plot below.

**Comment by forman** [comment]
I approve & merge, broken CI has other reasons.

**Pull request merged**